### PR TITLE
Invoking container GC immediately when a container has stopped

### PR DESCRIPTION
**Pull request by ronnielai**
- Support on-demand invocation of container GC 
- When PLEG detects a container has died, the container GC logic is invoked right away. Besides, the container GC logic gets invoked once per minute.

### CODE REVIEW

**Comment by ronnielai** [comment]
@vishh This PR is a simple mechanism to speed up the container GC by leveraging the container GC logic loop. Please let me know if a 5 second delay (at most) is tolerable. If not, I may have to use a more complex mechanism.

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -920,10 +926,20 @@ func (kl *Kubelet) listPodsFromDisk() ([]types.UID, error) {
 // Starts garbage collection threads.
 func (kl *Kubelet) StartGarbageCollection() {
 	go wait.Until(func() {
-		if err := kl.containerGC.GarbageCollect(kl.sourcesReady.AllReady()); err != nil {
-			glog.Errorf("Container garbage collection failed: %v", err)
+		kl.gcController.containerGCTrigger = make(chan struct{}, 1)
+		go wait.Until(func() {
```
Comment (vishh) [inline pkg/kubelet/kubelet.go]: nit: Add some comments describing the logic here and below.

**Comment by vishh** [comment]
This LGTM. @yujuhong can you take a quick look at this PR? It is trying to perform container GC whenever a container deletion is detected by kubelet.

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -920,10 +926,20 @@ func (kl *Kubelet) listPodsFromDisk() ([]types.UID, error) {
 // Starts garbage collection threads.
 func (kl *Kubelet) StartGarbageCollection() {
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: suggestion: can we start garbage collection in kubelet.Run(), so it's the same as other components?

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -920,10 +926,20 @@ func (kl *Kubelet) listPodsFromDisk() ([]types.UID, error) {
...
 	go wait.Until(func() {
-		if err := kl.containerGC.GarbageCollect(kl.sourcesReady.AllReady()); err != nil {
-			glog.Errorf("Container garbage collection failed: %v", err)
+		kl.gcController.containerGCTrigger = make(chan struct{}, 1)
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: suggestion: move gcController to a separate file and returns a channel at initialization

File: pkg/kubelet/kubelet.go [hunk 4]
```diff
@@ -920,10 +926,20 @@ func (kl *Kubelet) listPodsFromDisk() ([]types.UID, error) {
...
+		go wait.Until(func() {
+			kl.triggerContainerGarbageCollection()
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: This seems a little bit strange to me. Usually we just add a timer/ticker in the `select` below (line 935) for periodic tasks. No need to create a goroutine to trigger periodically.

**Comment by yujuhong** [comment]
> When PLEG detects a container has died, the container GC logic is invoked right away. Besides, the container GC logic gets invoked once per minute.

IIUC, this PR makes sure we clean up quickly after the container stopped. 
Do we need to worry about running GC too frequently? For example, deletion 20 containers will trigger 20 container in succession will trigger GC runs. How expensive is this and do we need to cap it or not?

**Comment by ronnielai** [comment]
If multiple containers are being deleted while container GC is going on, container GC will only run once after the current container GC finishes due to the channel buffer size being 1. So there's an implicit cap. However, it may be safer to have a cap on how often it gets invoked. @vishh, is a mandatory 5-second (suggestions welcome) gap between container GC runs acceptable?

**Comment by vishh** [comment]
Ideally, there should be no delays :P. Given that we get an hint on the
container that has stopped can we perform a light-weight GC that targets
only the pod that this container belongs to?

**Comment by ronnielai** [comment]
The Runetime interface doesn't provide container GC on a per pod basis...

**Comment by yujuhong** [comment]
> Ideally, there should be no delays :P. Given that we get an hint on the
> container that has stopped can we perform a light-weight GC that targets
> only the pod that this container belongs to?

Maybe. But if GC needs a global view of containers (or even a view of all containers in a pod), it'll run a `docker ps -a` anyway.  I'd rather not risk adding more load to docker unless we have a better way to avoid such calls. 
It should be easy to do a simple test to monitor how this change affects docker cpu usage during batch deletion of pods.

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (ronnielai) [inline pkg/kubelet/kubelet.go]: done

File: pkg/kubelet/kubelet.go (hunk 2 above)
Comment (ronnielai) [inline pkg/kubelet/kubelet.go]: StartGarbageCollection() and Run() are both methods in KubeletBootstrap interface. Changing how they interact seems to be outside the scope of this PR,

File: pkg/kubelet/kubelet.go (hunk 3 above)
Comment (ronnielai) [inline pkg/kubelet/kubelet.go]: done

File: pkg/kubelet/kubelet.go (hunk 4 above)
Comment (ronnielai) [inline pkg/kubelet/kubelet.go]: done

**Comment by ronnielai** [comment]
I've added a cap to enforce the minimum gap between container GC runs, which should avoid the potential thrashing. 

To avoid the docker ps call described in yujuhong's comment, we probably need to move to a model where there's no need to keep deleted containers around, which is tracked by #25326. I'll improve the container GC logic afterwards in a separate PR.

**Comment by Random-Liu** [comment]
> Maybe. But if GC needs a global view of containers (or even a view of all containers in a pod), it'll run a docker ps -a anyway. I'd rather not risk adding more load to docker unless we have a better way to avoid such calls. 
It should be easy to do a simple test to monitor how this change affects docker cpu usage during batch deletion of pods.

Why don't we use pod cache instead?

File: pkg/kubelet/kubelet.go [hunk 5]
```diff
@@ -923,10 +924,12 @@ func (kl *Kubelet) listPodsFromDisk() ([]types.UID, error) {
 // Starts garbage collection threads.
 func (kl *Kubelet) StartGarbageCollection() {
 	go wait.Until(func() {
-		if err := kl.containerGC.GarbageCollect(kl.sourcesReady.AllReady()); err != nil {
-			glog.Errorf("Container garbage collection failed: %v", err)
-		}
-	}, ContainerGCPeriod, wait.NeverStop)
+		kl.gcController.startContainerGarbageCollection(func() {
```
Comment (Random-Liu) [inline pkg/kubelet/kubelet.go]: @ronnielai Why do we need to start the garbage collector in a `wait.Until` clause? :)

File: pkg/kubelet/kubelet.go [hunk 6]
```diff
@@ -2551,6 +2554,9 @@ func (kl *Kubelet) syncLoopIteration(configCh <-chan kubetypes.PodUpdate, handle
 			glog.Errorf("Kubelet does not support snapshot update")
 		}
 	case e := <-plegCh:
+		if e.Type == pleg.ContainerDied {
+			kl.gcController.triggerContainerGarbageCollection()
+		}
```
Comment (Random-Liu) [inline pkg/kubelet/kubelet.go]: Do we also want to trigger garbage collection when new container started?

Comment (Random-Liu) [inline pkg/kubelet/kubelet.go]: NVM, the current gc policy doesn't care running containers.

**Comment by yujuhong** [comment]
> Why don't we use pod cache instead? Is there anything we can't get from the pod cache?

I believe PLEG sends out events before updating the pod cache. This will need to be changed if we want to use the pod cache for this specific case (i.e., processing containers). The garbage collection code will also need to be refactored to check only one pod.

The pod cache can't be used for the general logic yet because
 - It does not provide method to list all pods.
 - GC group all old containers created by kubelet (i.e., `unidentifiedContainers`) and delete them. This is not handled by the pod cache.

File: pkg/kubelet/kubelet.go (hunk 5 above)
Comment (ronnielai) [inline pkg/kubelet/kubelet.go]: Move the wait.Until inside startContainerGarbageCollection().

**Comment by Random-Liu** [comment]
@yujuhong Hm...Make sense to me. :)

> I believe PLEG sends out events before updating the pod cache. 
It seems that for each pod, PLEG updates pod cache first, and then sends out the event. However, for the whole cache, yeah when receiving the event, the pod cache may only be partially updated.

If so, maybe after pushing down the interface to container level, we can cache at container level. Then we can avoid some relist and share the container cache between pleg, gc and maybe other components. Just random thoughts. :)

**Comment by Random-Liu** [comment]
```
FAILED   ./hack/../hack/verify-boilerplate.sh	0s
```

**Comment by yujuhong** [comment]
s/Kubernetes Authors All rights reserved./he Kubernetes Authors.

**Comment by Random-Liu** [comment]
@yujuhong If we change the garbage collector to do per pod garbage collection, pod cache should work, right? :)

**Comment by yujuhong** [comment]
> @yujuhong If we change the garbage collector to do per pod garbage collection, pod cache should work, right? :)

Yes, with one exception I mentioned previously: "GC group all old containers created by kubelet (i.e., unidentifiedContainers) and delete them. This is not handled by the pod cache."

**Comment by ronnielai** [comment]
Thanks for the clarification. I'll work on per-pod container GC leveraging the pod cache. The existing periodic container GC will be left alone for now (the related flags are declared deprecated in 1.4), which takes care of unidentified containers.

Closing this PR
